pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -381,97 +382,108 @@ func (c *Controller) getTargetCaptureDevice(pc *crdv1alpha1.PacketCapture) strin
-	klog.V(2).InfoS("Started processing PacketCapture on the current Node", "name", pc.Name, "device", device)
-	go func() {
-		captureErr := c.performCapture(ctx, pc, pcState, device)
-		func() {
-			c.mutex.Lock()
-			defer c.mutex.Unlock()
-			c.numRunningCaptures -= 1
-			state := c.captures[pc.Name]
-			if state != nil {
-				state.phase = packetCapturePhaseComplete
-				state.err = captureErr
-			}
+func (c *Controller) startCapture(ctx context.Context, pc *crdv1alpha1.PacketCapture, state *packetCaptureState, device string) {
+	klog.InfoS("Starting packet capture on the current Node", "name", pc.Name, "device", device)
+	// Resync the PacketCapture on exit of the capture goroutine.
+	defer c.enqueuePacketCapture(pc)
+
+	var filePath string
+	var captureErr, uploadErr error
+	func() {
+		localFilePath := nameToPath(pc.Name)
+		file, err := getPacketFile(localFilePath)
+		if err != nil {
+			captureErr = err
+			return
+		}
+		defer file.Close()
 
-		}()
-		c.enqueuePacketCapture(pc)
+		var capturedAny bool
+		capturedAny, captureErr = c.performCapture(ctx, pc, state, file, device)
+		// If nothing is captured, no need to proceed.
+		if !capturedAny {
+			return
+		}
+		// If any is captured, upload it if required and update filePath in the status of the PacketCapture.
+		filePath = env.GetPodName() + ":" + localFilePath
+
+		if pc.Spec.FileServer == nil {
+			return
+		}
+		if uploadErr = c.uploadPackets(ctx, pc, file); uploadErr != nil {
+			return
+		}
+		filePath = fmt.Sprintf("%s/%s.pcapng", pc.Spec.FileServer.URL, pc.Name)
 	}()
-	return nil
+
+	if captureErr != nil {
+		klog.ErrorS(captureErr, "PacketCapture failed capturing packets", "name", pc.Name)
+	}
+	if uploadErr != nil {
+		klog.ErrorS(uploadErr, "PacketCapture failed uploading packets", "name", pc.Name)
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	state.phase = packetCapturePhaseComplete
+	state.filePath = filePath
+	state.captureErr = captureErr
+	state.uploadErr = uploadErr
+	c.numRunningCaptures -= 1
 }
 
 func (c *Controller) performCapture(
 	ctx context.Context,
 	pc *crdv1alpha1.PacketCapture,
 	captureState *packetCaptureState,
+	file afero.File,
 	device string,
-) error {
+) (bool, error) {
 	srcIP, dstIP, err := c.parseIPs(ctx, pc)
 	if err != nil {
-		return err
+		return false, err
 	}
-	pcapngFile, pcapngWriter, err := getPacketFileAndWriter(pc.Name)
+	pcapngWriter, err := pcapgo.NewNgWriter(file, layers.LinkTypeEthernet)
 	if err != nil {
-		return err
+		return false, fmt.Errorf("couldn't initialize a pcap writer: %w", err)
 	}
+	defer pcapngWriter.Flush()
 	updateRateLimiter := rate.NewLimiter(rate.Every(captureStatusUpdatePeriod), 1)
 	packets, err := c.captureInterface.Capture(ctx, device, srcIP, dstIP, pc.Spec.Packet)
 	if err != nil {
-		klog.ErrorS(err, "Failed to start capture")
-		return err
+		return false, err
 	}
-	klog.InfoS("Starting packet capture", "name", pc.Name, "device", device)
+	// Track any packet is captured.
+	capturedAny := false
```

Comment:
Maybe you can defined this to the returned value `(capturedAny bool, error) ` to make above return `false` to be clear as well.

Reply:
Named return values harm readability in long functions. I added comments to explain the function and the return values.